cpp/src/fil/infer.cu

```diff
@@ -72,10 +86,122 @@ __device__ __forceinline__ void infer_one_tree(tree_type tree, float* sdata,
     bool cond = isnan(val) ? !n.def_left() : val >= n.thresh();
     curr = n.left(curr) + cond;
   }
-  out[0] = tree[curr].output();
+  vec<1, output_type> out;
+  out[0] = tree[curr].base_node::output<output_type>();
+  return out;
 }
 
-template <int NITEMS, typename storage_type>
+template <int NITEMS,
+          leaf_value_t leaf_payload_type>  // = FLOAT_SCALAR
+struct tree_aggregator_t {
+  /** To compute accurately, would need to know the latest __CUDA_ARCH__
+      for which the code is compiled and which fits the SM being run on.
+      This is an approximation */
+  static const int ptx_arch = 750;
+  typedef cub::BlockReduce<vec<NITEMS, float>, FIL_TPB,
+                           cub::BLOCK_REDUCE_WARP_REDUCTIONS, 1, 1, ptx_arch>
+    BlockReduce;
+  typedef typename BlockReduce::TempStorage TempStorage;
+
+  vec<NITEMS, float> acc;
+  TempStorage* tmp_storage;
+
+  static size_t smem_finalize_footprint(int) { return sizeof(TempStorage); }
+  static size_t smem_accumulate_footprint(int) { return 0; }
+
+  __device__ __forceinline__ tree_aggregator_t(int, void* shared_workspace,
+                                               size_t)
+    : tmp_storage((TempStorage*)shared_workspace) {}
+  __device__ __forceinline__ void accumulate(
+    vec<NITEMS, float> single_tree_prediction) {
+    acc += single_tree_prediction;
+  }
+  __device__ __forceinline__ void finalize(float* out, int num_rows,
+                                           int output_stride) {
+    __syncthreads();
+    new (tmp_storage) TempStorage;
```

Comment:
No need to call the placement `new` here.

Reply:
do you mean `*tmp_storage = new TempStorage;` ?

Reply:
`TempStorage` shouldn't need explicit initialization.